FastLoader have a white list mode

rogermichoud opened the pull request:
A new API has been provided for the FastLoader to
selectively choose the streams it wants to load. It also
adds checkpoint streams for each stream added to the
whitelist.

Along, some refactoring of fastLoader testing.

corfu-repo commented on runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java:
```diff
@@ -90,6 +90,28 @@
+        this.streamsToLoad.addAll(streamsToLoad);
+
+        // Generate the streamsCheckpointId (we need to allow them as well)
+        streamsToLoad.forEach((streamId) ->
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the parentheses around the "streamId" parameter [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1611)

corfu-repo commented:
SonarQube analysis reported 2 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

corfu-repo commented on test/src/test/java/org/corfudb/recovery/Helpers.java:
```diff
@@ -0,0 +1,107 @@
+        ObjectsView.ObjectID mapId = new ObjectsView.
+                ObjectID(CorfuRuntime.getStreamID(streamName), type);
+
+        assertThat(!rt.getObjectsView().getObjectCache().containsKey(mapId));
```
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Complete the assertion. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2970)

bitmingw commented:
@rogermichoud We are having inconsistent interface for `addStreamToIgnore` and `addStreamsToLoad`. The ignore one takes table name and does a translate from table name to stream ID internally. I expect the add one to have a similar behavior: takes List<String> and do the transform.

rogermichoud commented:
@bitmingw Will do the change. Also, there is another issue I have to fix before you can test it.

corfudb-performance commented:
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **d079081**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #973 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/X8TfRZcHPlQ47SPZ3JYsZQO0Kjl6hsWt)

bitmingw commented:
@rogermichoud I think it basically works but ERROR logs are still printed out, which is annoying.

The problem can be described below.

```
loadMaps()
  - recoverRuntime()
    - processLogData()
      - updateCorfuObject()  <--- here
```

In `updateCorfuObject()` the first thing is constructing `logEntry` from `logData`. The deserailization will fail for a unknown type. Then the log `Cannot deserialize log entry` and its stack trace is printed out.

I purpose to suppress this ERROR message since it will confuse people.

rogermichoud commented:
I will fix it this way. There are 2 cases:

1. No stream in the logData (it can contain multiple streams), will just completely ignore the logData.
2. We have a mix of streams (some are in the whitelist, some are not), I will go ahead with the deserialization and log an error if we cannot deserialize it. This means that a client wrote a transaction with a mix of tables we are able to deserialize and some that we are not.

corfudb-performance commented:
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **ef8dad1**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #973 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/4yeIAA3pq7rDmcJMne09ESH2ahMx9Nju)

Maithem commented on runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java:
```diff
@@ -90,6 +90,26 @@
+     * the streams provided through this api.
+     * @param streamsToLoad
+     */
+    public FastObjectLoader addStreamsToLoad(List<String> streamsToLoad) {
```
What's the definition of white lists here, only load streams that are in the white list? 
Make sure that they are mutually exclusive (i.e. can't use both lists at the same time).

rogermichoud replied:
Will add in the comments that they are mutually exclusive. The way to enforce it is by using the whiteList flag.

Maithem commented on runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java:
```diff
@@ -232,20 +254,53 @@ public void updateStreamTails(long address, ILogData logData) {
      */
-    private boolean shouldEntryBeApplied(UUID streamId, SMREntry entry) {
+    private boolean shouldEntryBeApplied(UUID streamId, SMREntry entry, boolean isCheckpointEntry) {
+        // If white list mode, ignore everything that is not in the list or checkpoints
```
The comment is incorrect?

rogermichoud replied:
is poorly written.  I'll rephrase it. What I wanted to emphasize was that in this list there are both the streams passed by the client and their derived checkpoint streams.

Maithem commented on runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java:
```diff
@@ -232,20 +254,53 @@ public void updateStreamTails(long address, ILogData logData) {
-    private boolean shouldEntryBeApplied(UUID streamId, SMREntry entry) {
+    private boolean shouldEntryBeApplied(UUID streamId, SMREntry entry, boolean isCheckpointEntry) {
+        // If white list mode, ignore everything that is not in the list or checkpoints
+        if (whiteList && !streamsToLoad.contains(streamId)) {
```
This is kind of superfluous, only the second clause is needed. Also, why negate/negate?

If (streamsToLoad.contains(streamId)) return true;

rogermichoud replied:
The logic is a bit more complicated. The last condition should always be checked, whether we are in whitelist or blacklist. It checks that we are not processing an entry that is already contained in a checkpoint we already processed.

This is why if streamsToLoad.contains(streamId) return true is actually not enough. I had this at the beginning, and it was not passing the tests.

I will re-work this function to make it more self-explanatory. I reckon it is a bit hard to read.

Maithem commented on runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java:
```diff
@@ -232,20 +254,53 @@ public void updateStreamTails(long address, ILogData logData) {
+
         // We ignore the transaction stream ID because it is a raw stream
         // We don't want to create a Map for it.
         if (streamId == runtime.getObjectsView().TRANSACTION_STREAM_ID || streamsToIgnore.contains(streamId)) {
```
You can either add this stream to the black list, or not include it in the white list.

Maithem commented on runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java:
```diff
@@ -232,20 +254,53 @@ public void updateStreamTails(long address, ILogData logData) {
         }
 
-        if (!streamsMetaData.containsKey(streamId) || entry.getEntry().getGlobalAddress() >
+        if (isCheckpointEntry || !streamsMetaData.containsKey(streamId) || entry.getEntry().getGlobalAddress() >
```
Just use static boolean isCheckPointEntry(ILogData logData) ?

rogermichoud replied:
I am too deep for that. I don't have access to the logData directly (or I would need to get it backward from the logEntry). Here I am processing an entry inside the CheckpointEntry. Easiest way to know if we are processing a checkpoint or a normal entry is to pass it along the parameters.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java:
```diff
@@ -373,6 +373,14 @@ public static UUID getStreamID(String string) {
     }
 
+    public static UUID getCheckpointStreamIdFromId(UUID streamId) {
+        return getStreamID(streamId.toString() + "_cp");
```
Probably want to make the string lateral a static definition in streamview

rogermichoud replied:
sound right.

Maithem commented on test/src/test/java/org/corfudb/recovery/FastObjectLoaderTest.java:
```diff
@@ -813,233 +472,134 @@ public void canReadWithEntriesInterleavingCPS() throws Exception {
     }
 
     @Test
     public void ignoreStreamForRuntimeButNotStreamTails() throws Exception {
```
What happens when you white list and ignore a stream?

rogermichoud replied:
The expectation is that you should not be doing this. Modes are mutually exclusive, so adding a stream to ignore in the whitelist mode should be a prohibited operation. I added an exception when this happens.

Maithem commented:
Also, rebase.

rogermichoud commented on runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java:
```diff
@@ -90,6 +90,26 @@
     @Getter
     private boolean logHasNoCheckPoint = false;
 
+    private boolean whiteList = false;
```
I used it as the knob for the mutual exclusion between the whitelist mode and the blacklist mode. If the client of the API wants to do whitelist but because of its logic he doesn't have any table in the provided streamsToLoad, it should still be a whitelist. 

As is, I could not infer from the streasToLoad if we are in whitelist mode with no streams given or if we are in blacklist mode.

Maithem replied:
whiteList is set to true when you add a stream to a white list, so it can be inferred from the list size. Setting it to true even if the white list is empty is another issue.

Also, my initial point was its kind of redundant and it doesn't seem to work. Consider this example:

@Test
    public void whiteListMode() throws Exception {
        populateMaps(MORE, getDefaultRuntime(), CorfuTable.class, true, SOME);

        List<String> streamsToLoad = new ArrayList<>();
        streamsToLoad.add("Map1");
        streamsToLoad.add("Map3");

        CorfuRuntime rt2 = new CorfuRuntime(getDefaultConfigurationString()).connect();
        FastObjectLoader loader = new FastObjectLoader(rt2)
                .setDefaultObjectsType(CorfuTable.class);
        loader.addStreamToIgnore("a");
        loader.addStreamsToLoad(streamsToLoad);
        loader.loadMaps();

        assertThatMapsAreBuiltWhiteList(rt2, streamsToLoad);
    }

rogermichoud commented:
@Maithem I addressed the various comments. 

I agree it is debatable if the way I structured my testing is the best one, but I will let it like this for the 2 following reasons:
1. There is a lot of corner cases in these tests, refactoring them in bigger tests is kind of dangerous. I might miss some of these corner cases.
2. The advantage of having multiple small tests is that it is easier to understand what went wrong when some of them are not passing (and it happened many times). When refactoring the code, I saw that only the one related to checkpoints with extra data after them were failing. It was easier to locate where was the bug.

Beyond the style/small changes, I rework the shouldEntryBeApplied workflow to make it easier for the reader. Also added some test to ensure mutual exclusivity of whitelist/blacklist mode.

corfu-repo commented on runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java:
```diff
@@ -90,6 +100,32 @@
+     * @param streamsToLoad
+     */
+    public FastObjectLoader addStreamsToLoad(List<String> streamsToLoad) {
+        if (streamsToIgnore.size() != 0) {
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Use isEmpty() to check whether the collection is empty or not. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1155)

corfu-repo commented on runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java:
```diff
@@ -221,29 +265,85 @@ public void updateStreamTails(long address, ILogData logData) {
+     * @param entry entry under scrutinization.
+     * @return if the entry is already part of the checkpoint we started from.
+     */
+    private boolean entryAlreadyContainedInCheckpoint(UUID streamId, SMREntry entry) {
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "SMREntry"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

corfu-repo commented:
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

corfudb-performance commented:
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **b49359b**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #973 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/ojOnvJP7YQDRSstZAYAj6Sxh0f6AIb5Y)

rogermichoud commented:
@Maithem 
The two modes are mutually exclusive. The operation you mentioned is prohibited. These wrong usage of the loader (mixing modes) are tested in blackListAndWhiteListAreMutuallyExclusive.

Better be safe than sorry, I copied verbatim your example and it has the expected behavior, throwing a IllegalStateException: Cannot add a whitelist when there are already streams to ignore

Your concern on robustness and readibility is valid. The whiteMode flag actually reduce code in the addStreasToLoad method. Since an empty list is already initialized, I don't need to add extra logic to check if the list has been initialized (or use lombok lazy initialization that is sometimes a bit opaque).  I will argue that this is more code style than code functionality. Also, the alternative is to have a method isWhiteListModeEnable that infers from the list (and from some other mechanism to see if we passed a empty list) if we are in whiteListMode. I am not sure this would make anything more readable.

Anyway, as long as functionaly works and readibility is okay, I agree we should proceed.

Maithem commented:
Ok, I see the test failing now. Didn't see the force-push.
